# <a href="https://mipt-stats.gitlab.io/courses/ad_mipt.html">Phystech@DataScience</a>
## Задание 2

**Правила:**

* Выполненную работу нужно отправить телеграм-боту.
* Дедлайн **23 февраля в 15:00 (отбор)**. После дедлайна работы не принимаются кроме случаев наличия уважительной причины с подтверждающими документами.
* Прислать нужно ноутбук в формате `ipynb` 
* Решения, размещенные на каких-либо интернет-ресурсах, не принимаются. Публикация решения может быть приравнена к предоставлении возможности списать.
* Для выполнения задания используйте этот ноутбук в качествие основы, ничего не удаляя из него.

-----

**Работа Закировой Марфы, студентки третьего курса ФБМФ, Б06-907**

In [10]:
import numpy as np
import pandas as pd
from sklearn.datasets import load_diabetes
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LinearRegression
import matplotlib.pyplot as plt
import seaborn as sns
sns.set(font_scale=1.3)

Рассмотрим данные исследования 2004 года о зависимости риска развития диабета от пола, возраста, индекса массы тела, среднего кровяного давления и других показателей. 

Функция `sklearn.datasets.load_diabetes()` возвращает словарь. В поле `data` записана матрица регрессоров, в которой данные предварительно центрированы и нормированы. В поле `target` записана мера прогрессирования заболевания в течении года. В поле `DESCR` можно прочитать подробнее о данных.

### Загрузка данных

In [11]:
data = load_diabetes()

Посмотрим на описание датасета.

In [12]:
print(data['DESCR'])

.. _diabetes_dataset:

Diabetes dataset
----------------

Ten baseline variables, age, sex, body mass index, average blood
pressure, and six blood serum measurements were obtained for each of n =
442 diabetes patients, as well as the response of interest, a
quantitative measure of disease progression one year after baseline.

**Data Set Characteristics:**

  :Number of Instances: 442

  :Number of Attributes: First 10 columns are numeric predictive values

  :Target: Column 11 is a quantitative measure of disease progression one year after baseline

  :Attribute Information:
      - age     age in years
      - sex
      - bmi     body mass index
      - bp      average blood pressure
      - s1      tc, total serum cholesterol
      - s2      ldl, low-density lipoproteins
      - s3      hdl, high-density lipoproteins
      - s4      tch, total cholesterol / HDL
      - s5      ltg, possibly log of serum triglycerides level
      - s6      glu, blood sugar level

Note: Each of these 1

Поле `data` содержит матрицу размера 442 $\times$ 10, где 442 — количество пациентов, а 10 — количество признаков (возраст, пол, и т.д.). 
Строки матрицы соответствуют пациентам, столбцы — признакам.

In [13]:
data['data'].shape

(442, 10)

Целевая переменная $-$ мера прогрессирования заболевания в течении года.


In [14]:
data['target'].shape

(442,)

Создайте матрицу регрессоров $X$ и столбец наблюдений $y$.

In [ ]:
X, y = <...>

In [27]:
X, y = data['data'].copy(), data['target'].copy()

### Обучение моделей

Разбейте данные случайно на две части — обучающую и тестовую в соотношении 80:20.

Если что-то забыли или что-то не понятно, можно почитать справку:

In [16]:
train_test_split?

In [ ]:
X_train, X_test, y_train, y_test = <...>

print(X_train.shape, X_test.shape, y_train.shape, y_test.shape)

In [17]:
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2)

print(X_train.shape, X_test.shape, y_train.shape, y_test.shape)

(353, 10) (89, 10) (353,) (89,)


Заведите модель линейной регрессии из `sklearn` и обучите ее по обучающей части данных.

In [ ]:
model = <...>
<...>

In [18]:
from sklearn.linear_model import LinearRegression
model = LinearRegression()
model.fit(X_train, y_train)


LinearRegression()

Выведите кооэффициенты модели, включая свободный член.

In [19]:
print(f"Свободный член: {model.intercept_}")
print(f"Коэффициент: {model.coef_}")

Свободный член: 155.54980479186807
Коэффициент: [  -25.56817905  -241.87663834   497.55776227   354.22941929
 -1000.4983576    563.75663564   204.64117537   311.03928608
   776.18963273    89.87251412]


Посчитайте предсказания построенной модели на тестовой выборке

In [20]:
y_pred = model.predict(X_test)

Реализуйте метрики MSE, MAE, MAPE без использования `sklearn` и других готовых реализаций.

In [ ]:
def mean_squared_error(y_true, y_pred):
    <...>

def mean_absolute_error(y_true, y_pred):
    <...>

def mean_absolute_percentage_error(y_true, y_pred):
    <...>

In [22]:
def mean_squared_error(y_true, y_pred):
  return np.sum(np.power((y_true - y_pred),2))/(y_true.size)
  #каждое значение полученного при разности массива возводим в квадрат и суммируем все значения

def mean_absolute_error(y_true, y_pred):
  return np.sum(np.abs(y_true - y_pred))/(y_true.size)

def mean_absolute_percentage_error(y_true, y_pred):
  return 100 * np.sum(np.abs((y_true - y_pred) / y_true))/(y_true.size)


Посчитайте MSE, MAE, MAPE на тестовой выборке и выведите с точностью до трех знаков после запятой.

In [23]:
print(f"MSE: {np.around(mean_squared_error(y_test,y_pred),3)}")
print(f"MAE: {np.around(mean_absolute_error(y_test,y_pred),3)}")
print(f"MAPE: {np.around(mean_absolute_percentage_error(y_test, y_pred),3)}%")

MSE: 2882.619
MAE: 43.123
MAPE: 48.626%


###Своя реализация модели

Теперь реализуем линейную регрессию самостоятельно, используя формулы с лекции.
Вам нужно только заполнить прочерки в методах `fit` и `predict`.

In [ ]:
class MyLinearRegression:
    """
    Класс, реализующий линейную регрессию c помощью МНК.
    """
    
    def __init__(self):
        pass
        
    def fit(self, X, Y):
        """
        Функция обучения модели. 
        
        Предполагается модель Y = X * theta + epsilon. 
        где X --- регрессор (матрица размера n x d), 
        Y --- отклик (матрица размера n x 1),
        epsilon-ы имеют нормальное распределение

        Обратите внимание, здесь нет intercept_
        """
        
        self.n, self.d = <...>
        
        self.theta = <МНК-оценка>
  
        return self
    
        
    def predict(self, X):
        """
        Возвращает предсказание отклика на новых объектах X 
        и опционально доверительный/предсказательный интервал.
        
        X --- матрица объектов размера n x d
                        
        """
        
        y_pred = <...>
        return <...>

Заметим, что

In [29]:
np.linalg.det(np.dot(np.transpose(X), X)) != 0 

False

*Указанная матрица невырождена $\Rightarrow$ справедлива теорема по которой найдем значение $\widehat{\theta}$:*

$\widehat{\theta} = (X^T X)^{-1} X^T Y$

In [81]:
class MyLinearRegression:
    """
    Класс, реализующий линейную регрессию c помощью МНК.
    """
    
    def __init__(self):
        pass
        
    def fit(self, X, Y):
        """
        Функция обучения модели. 
        
        Предполагается модель Y = X * theta + epsilon. 
        где X --- регрессор (матрица размера n x d), 
        Y --- отклик (матрица размера n x 1),
        epsilon-ы имеют нормальное распределение

        Обратите внимание, здесь нет intercept_
        """
        
        self.n, self.d = X.shape[0], X.shape[1]  # X --- матрица объектов размера n x d
        theta_vspom = np.linalg.inv(np.dot(np.transpose(X), X))
        theta_vspom = np.dot(theta_vspom, np.transpose(X))
        self.theta = np.dot(theta_vspom, Y)
  
        return self
    
        
    def predict(self, X):
        """
        Возвращает предсказание отклика на новых объектах X 
        и опционально доверительный/предсказательный интервал.
        
        X --- матрица объектов размера n x d
                        
        """
        y_pred = np.dot(X, self.theta)
        return y_pred

Обучите вашу модель на данных о диабете с добавлением свободного члена.
Распечатайте коэффициенты и сравните их с коэффициентами модели из `sklearn`.

*Предсказание отклика на новом объекте x:*

$\widehat{Y} = X \widehat{\theta}$

Положим $x_1 == 1 \Rightarrow$ *регрессия со свободным членом $\theta_1$*

In [82]:
X_1_train = np.ones(X_train.shape[0])
X_train_new = np.column_stack((X_1_train, X_train))

my_model_svob = MyLinearRegression()
my_model_svob.fit(X_train_new, y_train)

print(f"Свободный член из my_model: {my_model_svob.theta[0]}")
print(f"Свободный член из sklearn: {model.intercept_}")
res_svob = pd.DataFrame({'Коэффициенты из my_model':my_model_svob.theta[1:] , 'Из sklearn': model.coef_})
res_svob

Свободный член из my_model: 155.54980479186813
Свободный член из sklearn: 155.54980479186807


Коэффициенты из my_model   Из sklearn
0                -25.568179   -25.568179
1               -241.876638  -241.876638
2                497.557762   497.557762
3                354.229419   354.229419
4              -1000.498358 -1000.498358
5                563.756636   563.756636
6                204.641175   204.641175
7                311.039286   311.039286
8                776.189633   776.189633
9                 89.872514    89.872514

Теперь обучите модель без свободного члена.

In [87]:
my_model = MyLinearRegression()
my_model.fit(X_train, y_train)


res = pd.DataFrame({'Коэффициенты из my_model':my_model.theta , 'sklearn': model.coef_})
res

Коэффициенты из my_model      sklearn
0                 69.345155   -25.568179
1               -223.635068  -241.876638
2                556.269641   497.557762
3                306.734262   354.229419
4               -970.985520 -1000.498358
5                410.106350   563.756636
6                279.106352   204.641175
7                494.010806   311.039286
8                773.180859   776.189633
9                -39.795907    89.872514

Сравните качество моделей со свободным членом и без него на тестовой выборке.

*Сравним по значениям метрик качества:*

In [86]:
y_pred_svob = my_model_svob.predict(np.column_stack((np.ones(X_test.shape[0]), X_test)))
y_pred_ = my_model.predict(X_test)

metrics = pd.DataFrame({"Cо свободным членом":[np.around(mean_squared_error(y_test, y_pred_svob), 3), 
                                               np.around(mean_absolute_error(y_test, y_pred_svob), 3),
                                               np.around(mean_absolute_percentage_error(y_test, y_pred_svob), 3)],
                       "Без него": [np.around(mean_squared_error(y_test, y_pred_), 3), 
                                    np.around(mean_absolute_error(y_test, y_pred_), 3), 
                                    np.around(mean_absolute_percentage_error(y_test, y_pred_), 3)]})
metrics.index = ["MSE", "MAE", "MAPE"]
metrics

Cо свободным членом   Без него
MSE              2882.619  22280.887
MAE                43.123    139.942
MAPE               48.626    122.859

Сделайте выводы.

*Данные по собственной модели со свободным членом и sklearn можно считать одинаковыми. При этом при попытке приблизить результаты линейной регрессией без свободного члена получаем большую ошибку, по данным метрик качества получаем ухудшение в разы.*